### ISSUE TITLE
[bug] Nuitka produces runtime TypeError with Pydantic's BaseModel

### ISSUE THREAD

**Issue by jborman-exos** [Nuitka/Nuitka#1359]
A module utilizing Pydantic's `BaseModel` class is compiled via Nuitka. The plain source code runs fine when executed in `test_my_module.py`, however the compiled code fails to pass. The resulting error when trying to invoke `b.bar(bar_arg='some argument')` causes the following exception. (Code is included down below in the body of the issue template).

```
Traceback (most recent call last):
  File "/home/[user1]/.pyenv/versions/3.9.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/[user1]/.pyenv/versions/3.9.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/mnt/c/Users/[user1]/github/nuitka-pydantic-bug/test_my_module.py", line 19, in <module>
    test_minimum_failure()
  File "/mnt/c/Users/[user1]/github/nuitka-pydantic-bug/test_my_module.py", line 8, in test_minimum_failure
    b.bar(bar_arg='some argument')
TypeError: bar() missing 1 required positional argument: 'self'
```
---

* Nuitka version, full Python version, flavor, OS, etc. as output by this
  command (it does more than you think, and we are adding more all the time):

  > python -m nuitka --version

```
(.venv) [user1]@LAPTOP:/mnt/c/Users/[user1]/github/nuitka-pydantic-bug$ python -m nuitka --version
0.6.18.6
Commercial: None
Python: 3.9.9 (main, Dec 29 2021, 12:20:17)
Flavor: pyenv
Executable: /mnt/c/Users/[user1]/github/nuitka-pydantic-bug/.venv/bin/python
OS: Linux
Arch: x86_64
Distribution: Ubuntu "16.04.6
```

* How did you install Nuitka and Python

```
Installed via pip into a virtualenv (WSL Ubuntu). Also tested on plain Windows with pip + venv. Confirming that Python.exe is not from the Windows store.
```

* The specific PyPI names and versions

  It should be taken from this output if there specific packages involved, e.g.
  `numpy`, you are expected to shorten this to the relevant ones.

  > python -m pip freeze

```
Nuitka==0.6.18.6
pydantic==1.9.0
typing_extensions==4.0.1
```

* Many times when you get an error from Nuitka, your setup may be special

  Then even a `print("hello world")` program will not work, please try that and report
  that instead.

```
Confirming simple "hello world" works
```

* Also supply a [Short, Self Contained, Correct, Example](https://sscce.org/)

 See [here](https://github.com/jborman-exos/nuitka-pydantic-bug) for the repo with the example below.

**my_module/__init__.py**:

``` python
from pydantic import BaseModel

class A:
    def foo(self, foo_arg):
        print(f'A.foo({foo_arg})')

class B(BaseModel):
    member_a: A

    def bar(self, bar_arg: str) -> None:
        self.member_a.foo(bar_arg)

    class Config:
        arbitrary_types_allowed = True


class B2:
    def __init__(self, member_a: A) -> None:
        self.member_a = member_a

    def bar(self, bar_arg: str) -> None:
        self.member_a.foo(bar_arg)
```

**test_my_module.py**:

``` python
import my_module as mm

def test_minimum_failure() -> None:
    a = mm.A()
    b = mm.B(member_a=a)
    print(f'b: {b}', type(b))
    print(f'b.member_a: {b.member_a}', type(b.member_a))
    b.bar(bar_arg='some argument')

def test_working() -> None:
    a = mm.A()
    b2 = mm.B2(member_a=a)
    print(f'b2: {b2}', type(b2))
    print(f'b2.member_a: {b2.member_a}', type(b2.member_a))
    b2.bar(bar_arg='some argument')

if __name__ == '__main__':
    test_working()
    test_minimum_failure()
```

* Provide in your issue the Nuitka options used

```
python -m nuitka --module my_module --include-package my_module
```

* Note if this is a regression

  If it used to work in an earlier version of Nuitka, please note what you know
  about that.

```
I have tried the following versions: `0.6.19rc8`, `0.6.18.6` and `0.6.15.3`
```

**Comment by jborman-exos** [comment]
To expand on the above, the exact issue is that an instance of class `B` fails to execute `b.bar(...)`without passing in a reference to itself:

``` python
import my_module as mm

a = mm.A()
b = mm.B(member_a=a)

b.bar('some arg')    # raises `TypeError: missing positional argument 'bar_arg'`
b.bar(bar_arg='some arg')  # raises `TypeError: missing positional argument 'self'`
b.bar(b, 'some arg')    # Ok
```

Executing `b.bar(b, 'some arg')` with a reference to itself works as (un)expected...

It seems like a compiled class inheriting from `BaseModel` does not pass a `self` reference to the member's method. I'm not familiar with what Nuitka is doing under the hood, but hopefully all this info is helpful.

**Comment by kayhayen** [comment]
I am suspecting this is #471 which has a solution, check it out, will need 0.6.19 to be really good, but 0.6.18 series has something for it, but I definitely recommend to use develop for now. The 0.6.19 release is due soon.

**Comment by jborman-exos** [comment]
Confirming that using nuitka `0.6.19rc9` with `--experimental=function-base` resolves the runtime error

**Comment by kayhayen** [comment]
You are spot on, to ask, this is indeed the solution, it's just that with the hotfix, some things work with it, and now all the things are supposed to work, and in fact it could be the baseline code if I was more daring. However I will probably want to make it on a develop branch with 0.6.20rc to gather feedback if that is really good. I am very conservative with Nuitka.

**Comment by hjzCy** [comment]
nuitka `v0.6.19` using `--clang` , `--experimental=function-base` option and `vs2022` compile error:

```shell
C:\Users\78081\AppData\Local\pypoetry\Cache\VIRTUA~1\KFCLIE~2.8\lib\SITE-P~1\nuitka\build\static_src\CompiledFrameType.c(544,5): error: initializer element is not a compile-time constant
    &PyFrame_Type, // tp_base
    ^~~~~~~~~~~~~
4 errors generated.
Backend C: 0%| | 0/556s
cons: *** [static_src\CompiledGeneratorType.obj] Error 1
static_src\CompiledFunctionType.c(523,5): error: initializer element is not a compile-time constant
    &PyFunction_Type, /* tp_base */
    ^~~~~~~~~~~~~~~~~~~
In file included from static_src\CompiledFunctionType.c:2722:
C:\Users\78081\AppData\Local\pypoetry\Cache\VIRTUA~1\KFCLIE~2.8\lib\SITE-P~1\nuitka\build\static_src\CompiledMethodType.c(526,5): error: initializer element is not a compile-time constant
    &PyMethod_Type, /* tp_base */
    ^~~~~~~~~~~~~~~~
2 errors generated.
Backend C: 0%| | 0/556s
cons: *** [static_src\CompiledFunctionType.obj] Error 1
````

clang component is: `v143 build tool C++ Clang (x64/x86)`, `Windows for C++ Clang compiler (12.0.0)`
python version: `v3.8.10 (x64)`
pydantic: `v1.9.0`
typing_extensions: `v4.0.1`

If you do not inherit `BaseModel`, you can compile successfully without the `--experimental=function-base` option.

**Comment by kayhayen** [comment]
Ah, clangcl @hjzCy I might fix that up in a hotifx coming to 0.6.19, only MSVC proper seems to compile it at this time.

**Comment by kayhayen** [comment]
So, on factory and on coming hotfix, this is compiling just fine now: https://nuitka.net/doc/factory.html

**Comment by kayhayen** [comment]
Note, I am soon going to make this the default. I noticed it's missing checks for the method type however, I will want to add those first.

**Comment by murphp15** [comment]
Hey Guys, 
I am using the factory build of nuitka and still getting runtime issues. 

Stacktrace: 
```
  File "pydantic/main.py", line 228, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 488, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 419, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 528, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 559, in pydantic.fields.ModelField._set_default_and_type
pydantic.errors.ConfigError: unable to infer type for attribute "cpu"
```

the variable looks like 

```
    cpu: Optional[bool] = None
```

pyproject.toml
```
[build-system]
requires = ["setuptools", "wheel", "setuptools_scm>=6.2", "https://github.com/Nuitka/Nuitka/archive/factory.zip"]
build-backend = "nuitka.distutils.Build"
```

setup.py
```
setup(
    packages=find_packages('src'),
    package_dir={'': 'src'},
    python_requires=">=3.8",
    install_requires=load_requirements(),
    extras_require={
        "dev": load_requirements("requirements-dev.txt"),
    },
    command_options={
        'nuitka': {
            '--experimental': 'function-base'
        }
    },
)
```


Do you have any idea how I might address this?

**Comment by kayhayen** [comment]
Looks good to me in terms of options @murphp15 not sure what's going on there, either the arguments are not passed correctly, or pydantic has another issue, can you provide a minimal reproducer without setuptools and much software used?

**Comment by kayhayen** [comment]
And also, if you have `--show-scons` in your options, it would show the command being executed, just to make sure the proper command line to Nuitka gets composed.

**Comment by murphp15** [comment]
@kayhayen Is there files that we should be looking for in the build to indicate that this infer information has been kept? 
Or is there a way we can explicitly add this info for pydantic?

**Comment by kayhayen** [comment]
Looks good to me @murphp15 so this might be another issue, unfortunately. If you wanted an NDA, consider getting in touch via https://nuitka.net/doc/commercial.html, and I could work on this within your environment. But without a reproducer, there is nothing I can do really except spending time on trying out random things, which is not how I can spend my time.

**Comment by kayhayen** [comment]
There were changes necessary for 3.10 to not crash with generators, it is currently the default on the factory branch and seems proven, but currently is held back by the other changes there not being yet entirely ready for release.

**Comment by kayhayen** [comment]
This is part of the 0.7rc1 pre-release on develop right now.

**Comment by kayhayen** [comment]
This is part of the 0.7 release now.
